tools/runner/reporter.go

```diff
@@ -0,0 +1,102 @@
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package runner contains code for a test runner that can run a list of
+// load tests, wait for them to complete, and report on the results.
+package runner
+
+import (
+	"fmt"
+	"log"
+	"time"
+
+	grpcv1 "github.com/grpc/test-infra/api/v1"
+)
+
+// TestSuiteReporter manages reports for tests that share a runner queue.
+type TestSuiteReporter struct {
+	qName         string
+	logPrefixFmt  string
+	testCaseCount int
+}
+
+// NewTestSuiteReporter creates a new suite reporter instance.
+func NewTestSuiteReporter(qName string, logPrefixFmt string) *TestSuiteReporter {
+	return &TestSuiteReporter{
+		qName:        qName,
+		logPrefixFmt: logPrefixFmt,
+	}
+}
+
+// Queue returns the name of the queue containing tests for this test suite.
+func (r *TestSuiteReporter) Queue() string {
+	return r.qName
+}
+
+// NewTestCaseReporter creates a new reporter instance.
+func (r *TestSuiteReporter) NewTestCaseReporter(config *grpcv1.LoadTest) *TestCaseReporter {
+	logPrefix := fmt.Sprintf(r.logPrefixFmt, r.qName, r.testCaseCount)
+	r.testCaseCount++
+	return &TestCaseReporter{
+		logPrintf: func(format string, v ...interface{}) {
+			log.Printf(logPrefix+format, v...)
+		},
+	}
+}
+
+// TestCaseReporter collects events for logging and reporting during a test.
+type TestCaseReporter struct {
+	StartTime time.Time
+	Duration  time.Duration
+	logPrintf func(format string, v ...interface{})
+	index     int
+}
+
+// Index returns the index of the test case in the test suite (and queue).
+func (r *TestCaseReporter) Index() int {
+	return r.index
+}
+
+// Info records an informational message generated by the test.
+func (r *TestCaseReporter) Info(format string, v ...interface{}) {
+	r.logPrintf(format, v...)
+}
+
+// Warning records a warning message generated during the test.
+// The error that caused the message to be generated is also included.
+func (r *TestCaseReporter) Warning(format string, v ...interface{}) {
+	// TODO: Record warning.
+	r.logPrintf(format, v...)
+}
+
+// Error records an error message generated during the test.
+// The error that caused the message to be generated is also included.
+func (r *TestCaseReporter) Error(format string, v ...interface{}) {
+	// TODO: Record error.
+	r.logPrintf(format, v...)
+}
+
+// StartTest records the start of the test and its start time.
+func (r *TestCaseReporter) StartTest(startTime time.Time) {
+	r.StartTime = startTime
+}
+
+// EndTest records the end of the test and its end time.
+func (r *TestCaseReporter) EndTest(endTime time.Time) {
+	r.Duration = endTime.Sub(r.StartTime)
+}
```

Comment:
nit: Perhaps, do we need to inject the times? These functions feel like setters for attributes which are already exposed on the struct.

From my perspective, several assumptions are made:

- When the name is StartTest, an assumption is that the test starts with the call. For this reason, I think we can use `r.StartTime = time.Now()`.

- When the name is EndTest, an assumption is likewise made that the test ends with the call. To simplify the logic here, we can replace `r.Duration = time.Since(r.StartTime)`.

If the goal is to mock time for testing, there are some other ways we can achieve that. We can define a helper and override it to set the time to the proper point. That said, I think the majority of tests for the reporter would be validating the XML fields, not strings.